core/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -78,27 +78,36 @@
     private final AtomicBoolean finished = new AtomicBoolean();
 
     private final ConcurrentMap<String, IndexOutput> openIndexOutputs = ConcurrentCollections.newConcurrentMap();
-    private final CancellableThreads cancellableThreads = new CancellableThreads();
+    private final CancellableThreads cancellableThreads;
 
     // last time this status was accessed
     private volatile long lastAccessTime = System.nanoTime();
 
     private final Map<String, String> tempFileNames = ConcurrentCollections.newConcurrentMap();
 
+    public RecoveryTarget(RecoveryTarget copyFrom) { // copy constructor
+        this(copyFrom.indexShard(), copyFrom.sourceNode(), copyFrom.listener, copyFrom.cancellableThreads, copyFrom.recoveryId());
+    }
+
     public RecoveryTarget(IndexShard indexShard, DiscoveryNode sourceNode, RecoveryTargetService.RecoveryListener listener) {
+        this(indexShard, sourceNode, listener, new CancellableThreads(), idGenerator.incrementAndGet());
+    }
 
+    private RecoveryTarget(IndexShard indexShard, DiscoveryNode sourceNode, RecoveryTargetService.RecoveryListener listener,
+                           CancellableThreads cancellableThreads, long recoveryId) {
         super("recovery_status");
-        this.recoveryId = idGenerator.incrementAndGet();
+        this.cancellableThreads = cancellableThreads;
+        this.recoveryId = recoveryId;
         this.listener = listener;
         this.logger = Loggers.getLogger(getClass(), indexShard.indexSettings().getSettings(), indexShard.shardId());
         this.indexShard = indexShard;
         this.sourceNode = sourceNode;
         this.shardId = indexShard.shardId();
         this.tempFilePrefix = RECOVERY_PREFIX + indexShard.recoveryState().getTimer().startTime() + ".";
```

Comment:
hmm.. wondering if we should add a unique generation to these files (or sample the idGenerator again when we restart). Right now we will have collisions between temp names of the previous recovery target and the new one, but I think we want complete separation?